docs/HowToGuides/RunningIncludeWhatYouUse.md

```diff
@@ -0,0 +1,289 @@
+    `CMAKE_EXPORT_COMPILATION_COMMANDS=ON`. If you use `build-script` to
+    manage your builds, you can do this with
+    ```
+    swift/utils/build-script <other options> \
+      --extra-cmake-options='-DCMAKE_EXPORT_COMPILATION_COMMANDS=ON'
+    ```
+- Install [`jq`](https://stedolan.github.io/jq/). It's not strictly necessary,
+  but we will use it for some JSON munging.
+
+## Cloning and branch checkout
+
+The directory structure we will be using is
+
+```
+swift-project/
+  |--- build/
+  |     |--- Ninja-ReleaseAssert/
+  |     |     |--- compile_commands.json
+  |     |     `--- ...
+  |     `--- ...
+  |--- swift/
+  |--- iwyu/
+  |     |--- build/
+  |     |--- include-what-you-use/ # source
+  |     |--- logs/
+  |     `--- scripts/
+  `--- ...
+```
+
+Start out with `swift-project` as the working directory.
+
+1. Check out IWYU.
+   ```
+   mkdir iwyu
+   cd iwyu
+   git clone https://github.com/include-what-you-use/include-what-you-use.git
+   ```
+2. Find out the version of the `clang` built recently.
+   ```
+   ../build/Ninja-ReleaseAssert/llvm-macosx-x86_64/bin/clang --version
+   ```
+   This should say something like `clang version 10.0.0` or similar.
+3. Based on the `clang` version, make sure you check out the correct branch.
+   ```
+   cd include-what-you-use
+   git checkout clang_10
+   cd ..
+   ```
+
+## Building IWYU
+
+We are now in the `iwyu` directory
+
+1. Configure IWYU with CMake.
+   ```
+   mkdir build
+   cd build
+   cmake -G Ninja \
+     -DCMAKE_PREFIX_PATH=../../build/Ninja-ReleaseAssert/llvm-macosx-x86_64 \
+     -DCMAKE_CXX_STANDARD=14 \
+     ../include-what-you-use
+   ```
+2. Build IWYU
+   ```
+   cmake --build .
+   ```
+3. Create the `lib` directory if missing.
+   ```
+   file lib || ln -s ../../build/Ninja-ReleaseAssert/llvm-macosx-x86_64/lib lib
+   ```
+   This is needed for IWYU to find necessary Clang headers.
+4. Spot check IWYU for a basic C example.
+   ```
+   echo '#include <stdint.h>' > tmp.c
+   ./bin/include-what-you-use tmp.c -E -o /dev/null \
+     -I "$(xcrun --show-sdk-path)/usr/include"
+   rm tmp.c
+   ```
+   You should see output like:
+   ```
+   tmp.c should add these lines:
+
+   tmp.c should remove these lines:
+   - #include <stdint.h>  // lines 1-1
+
+   The full include-list for tmp.c:
+   ---
+   ```
+5. Spot check IWYU for a basic C++ example. Notice the extra C++-specific
+   include path.
+   ```
+   echo '#include <string>\n#include <cmath>' > tmp.cpp
+   ./bin/include-what-you-use tmp.cpp -E -o /dev/null \
+     -I "$(clang++ -print-resource-dir)/../../../include/c++/v1" \
+     -I "$(xcrun --show-sdk-path)/usr/include"
+   rm tmp.cpp
+   ```
+   You should see output like:
+   ```
+   tmp.cpp should add these lines:
+
+   tmp.cpp should remove these lines:
+   - #include <cmath>  // lines 2-2
+   - #include <string>  // lines 1-1
+
+   The full include-list for tmp.cpp:
+   ---
+   ```
+6. Leave the build directory.
+   ```
+   cd ..
+   ```
+
+## Running IWYU
```

Comment:
> apt-get install iwyu

The problem with suggesting this is that it doesn't provide any guarantees of what version of Clang `iwyu` will have been compiled against, so I'm hesitant to recommend it. For example, if that Clang doesn't recognize attributes that are used by Swift, maybe that's a problem?

Reply:
Fair enough.  Getting at least a version of the documentation in is still a good step forward :)

Reply:
>  I don't know about the version mismatch, I expect that the attributes shouldn't matter to IWYU as they are provided by the compiler, not headers.

Right, there's also the concern about the Clang simply being too old and having bugs etc. Recently there was some issue with the Ubuntu 16.04 bot failing because the Clang in it is too out-dated. Since it's not too difficult to build IWYU from source, given that we already have a Clang build available, I thought it's probably best to recommend that and reduce the number of moving parts.

Reply:
Hmm, I actually had issues with building it; but, seems reasonable to just build it against the local clang.